frontend/src/lib/components/SaveToDashboard/saveToDashboardModalLogic.ts

```diff
@@ -47,9 +47,37 @@ export const saveToDashboardModalLogic = kea<saveToDashboardModalLogicType<SaveT
         _dashboardId: [null as null | number, { setDashboardId: (_, { id }) => id }],
         searchQuery: ['', { setSearchQuery: (_, { query }) => query }],
         scrollIndex: [-1 as number, { setScrollIndex: (_, { index }) => index }],
+        adding: [false, { addToDashboard: () => true, reportSavedInsightToDashboard: () => false }],
+        removing: [
+            false,
+            {
+                removeFromDashboard: () => true,
+                reportRemovedInsightFromDashboard: () => false,
+            },
+        ],
+        dashboardWithActiveAPICall: [
+            null as number | null,
+            {
+                addToDashboard: (_, { dashboardId }) => dashboardId,
+                removeFromDashboard: (_, { dashboardId }) => dashboardId,
+                reportSavedInsightToDashboard: () => null,
+                reportRemovedInsightFromDashboard: () => null,
+            },
+        ],
     },
 
     selectors: {
+        buttonLabel: [
+            (s) => [s.adding, s.dashboardWithActiveAPICall],
+            (adding: boolean, dashboardWithActiveAPICall: number | null) =>
+                (dashboardId: number, isAlreadyOnDashboard: boolean) => {
+                    if (dashboardWithActiveAPICall === dashboardId) {
+                        return adding ? 'Adding' : 'Removing'
+                    } else {
+                        return isAlreadyOnDashboard ? 'Added' : 'Add to dashboard'
+                    }
+                },
+        ],
```

Comment:
Can't we derive whether the adding/removing action is being performed from `isAlreadyOnDashboard`?

Reply:
Doh!